rir/src/compiler/pir/instruction.h

```diff
@@ -279,6 +280,9 @@ class InstructionImplementation : public Instruction {
     static constexpr bool mayLeakEnv_ = ENV >= EnvAccess::Leak;
 
     bool hasEffect() const override { return EFFECT > Effect::None; }
+    bool hasObservableEffect() const override {
+        return EFFECT > Effect::Order && hasEffect();
```

Comment:
some subclasses override `hasEffect`. Since there is an ordering in the effects, if an instruction has no effect at all, it also doesn't have an observable effect....